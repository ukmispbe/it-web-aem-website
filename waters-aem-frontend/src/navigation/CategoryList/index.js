import React from "react";
import PropTypes from 'prop-types';
import ReactSVG from 'react-svg';
import { elementLocator } from "../../utils/eCommerceFunctions.js";

const CategoryList = ({items, activeIndex, onClick}) => {
    const categoryRef = React.useRef();
    return (
        <div className="cmp-category-wrapper">
            <div><h3>Category</h3></div>
            <div ref={categoryRef} className="cmp-category-items">
                {items.map((item, index) => 
                {
                    let backImage;
                    if (index === 0) {
                        backImage =  <ReactSVG className="cmp-categories-back" src="/content/dam/waters/en/brand-assets/icons/multiple.svg" />;
                    }
                    const isHidden = determineIfHidden(items, index, activeIndex);

                    return <Category key={`Category-${index}`} name={item.translation} count={item.count} index={index} isActive={index === activeIndex} onClick={onClick}  backImage={backImage} isHidden={isHidden} />
                }
                )}
            </div>
            {activeIndex !== 0 && <hr className="cmp-category-separator h-large" />}
        </div>
    );
}

const determineIfHidden = (items, index, activeIndex) => {
    if (activeIndex === 0) {
        if (index === 0) {
            return true;
        }
        return false;
    } 
    else {
        if (activeIndex === index || index === 0) {
            return false;
        }
        return true;
    }
}

const Category = ({index, name, count, isActive, onClick, backImage, isHidden}) => 
    <div className={`cmp-category-item${isActive ? " active" : ""} ${isHidden ? " hidden" : ""}`} onClick={() => onClick(index)}>
        {backImage}
        <span className="cmp-category-label" data-locator={elementLocator(name)}>{name}</span>
        <span className="cmp-category-count" data-locator={elementLocator(count)}> ({count})</span>
    </div>;


Category.propTypes = {
    name: PropTypes.string.isRequired,
    index: PropTypes.number.isRequired,
    isActive: PropTypes.bool.isRequired,
    onClick: PropTypes.func.isRequired,
    isHidden: PropTypes.bool.isRequired
};

Category.defaultProps = {
    name: "",
    index: -1,
    isActive: false,
    onClick: () => {},
    isHidden: false
}

CategoryList.propTypes = {
    items: PropTypes.array.isRequired,
    activeIndex: PropTypes.number.isRequired,
    onClick: PropTypes.func.isRequired
}

CategoryList.defaultProps = {
    items: [],
    activeIndex: -1,
    onClick: () => {}
}
export default CategoryList;