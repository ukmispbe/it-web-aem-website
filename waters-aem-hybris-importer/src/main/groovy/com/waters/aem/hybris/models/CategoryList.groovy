package com.waters.aem.hybris.models

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonFormat.Shape
import groovy.transform.ToString

@ToString(includePackage = false, includeNames = true)
class CategoryList {

    String id

    @JsonFormat(shape = Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss'Z'")
    Calendar lastModified

    List<Category> categories
}
