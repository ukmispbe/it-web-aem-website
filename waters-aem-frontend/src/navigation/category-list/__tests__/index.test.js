jest.mock("../../../scripts/fade-x.js");

import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import CategoryList from "..";

describe("Feature: CategoryList Component", () => {
    const propsWithOutItems = {
        text: {
            categoryText: "Category"
        },
        items: [],
        activeIndex: -1,
        onClick: jest.fn()
    };

    const propsWithItems = {
        ...propsWithOutItems,
        activeIndex: 0,
        items: [
            {
                count: 1525,
                name: "Library",
                translation: "Library"
            },
            {
                count: 8243,
                name: "Shop",
                translation: "Shop"
            }
        ]
    };
    
    beforeEach(() => {
        React.useEffect = jest.fn(f => f());
    });

    afterEach(() => {
        jest.restoreAllMocks();
    });

    describe("Scenario: Rendering CategoryList", () => {
        describe("When no categories are provided", () => {
            it("Then it should match snapshot", () => {
                const json = renderer.create(<CategoryList {...propsWithOutItems} />);
                
                expect(json).toMatchSnapshot();
            });
        });

        describe("When categories are provided", () => {
            it("Then it should match snapshot", () => {
                const json = renderer.create(<CategoryList {...propsWithItems} />);
                
                expect(json).toMatchSnapshot();
            });
        });
    });

    describe("Scenario: User Interaction", () => {
        describe("When Category is clicked", () => {
            it("Then it should call the click handler property", () => {
                const wrapper = shallow(<CategoryList {...propsWithItems} />);

                wrapper.find("Category").first().simulate("click");

                expect(propsWithItems.onClick).toHaveBeenCalled();
            });
        });
    });
});