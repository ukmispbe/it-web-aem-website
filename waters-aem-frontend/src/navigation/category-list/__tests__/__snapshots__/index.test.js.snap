// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feature: CategoryList Component Scenario: Rendering CategoryList When categories are provided Then it should match snapshot 1`] = `
<div
  className="cmp-category-wrapper"
>
  <div>
    <h3>
      Category
    </h3>
  </div>
  <div
    className="cmp-category-items"
  >
    <div
      className="cmp-category-item active "
      onClick={[Function]}
    >
      <span
        className="cmp-category-label"
        data-locator="library"
      >
        Library
      </span>
      <span
        className="cmp-category-count "
        data-locator={1525}
      >
         (
        1525
        )
      </span>
    </div>
    <div
      className="cmp-category-item "
      onClick={[Function]}
    >
      <span
        className="cmp-category-label"
        data-locator="shop"
      >
        Shop
      </span>
      <span
        className="cmp-category-count "
        data-locator={8243}
      >
         (
        8243
        )
      </span>
    </div>
  </div>
  <hr
    className="cmp-category-separator h-large"
  />
</div>
`;

exports[`Feature: CategoryList Component Scenario: Rendering CategoryList When no categories are provided Then it should match snapshot 1`] = `
<div
  className="cmp-category-wrapper"
>
  <div>
    <h3>
      Category
    </h3>
  </div>
  <div
    className="cmp-category-items"
  />
  <hr
    className="cmp-category-separator h-large"
  />
</div>
`;
